File: 프로그래머스/2/87946. 피로도/피로도.java
class Solution {
    static int result = 0;
    
    public int solution(int k, int[][] dungeons) {
        boolean[] visited = new boolean[dungeons.length];
        
        dfs(dungeons, visited, 0, k);
        
        return result;
    }
    
    public static void dfs(int[][] dungeons, boolean[] visited, int cnt, int k) {
        result = Math.max(result, cnt);
        
        for(int i=0; i<dungeons.length; i++) {
            if(!visited[i] && dungeons[i][0] <= k) {
                visited[i] = true;
                
                dfs(dungeons, visited, cnt + 1, k - dungeons[i][1]);
                
                visited[i] = false;
            }
        }
    }
}

File: 프로그래머스/2/87946. 피로도/피로도.py
result = 0

def solution(k, dungeons):
    visited = [False for _ in range(len(dungeons))]

    dfs(k, dungeons, visited, result)

    return result

def dfs(k, dungeons, visited, cnt):
    global result
    result = max(cnt, result)

    for i, dungeon in enumerate(dungeons):
        if not visited[i] and k >= dungeon[0]:
            visited[i] = True

            dfs(k - dungeon[1], dungeons, visited, cnt + 1)

            visited[i] = False

    return